### DISCUSSION TITLE
The Case for Crashpad

### DISCUSSION THREAD

**Discussion by Nerixyz** [Chatterino/chatterino2#4349] [Ideas]
There are some [open issues](https://github.com/Chatterino/chatterino2/issues?q=crash+is%3Aopen) regarding Chatterino crashing. There might be a lot more crashes happening "in the wild" than the ones reported (e.g. one-time crashes where the user just restarts the app). Chatterino previously had an option to add [breakpad]. This was never used in CI builds, thus the majority of users never had the ability to report crashdumps. For the majority of users, it's not viable to run Chatterino in debug mode or with a debugger attached to properly document crashes.

I'm proposing the addition of [crashpad]. This doesn't need to be integrated into every build. It should be enough to have one build that includes crashpad for every release. I'm leaning towards crashpad over breakpad, because it [seems](https://docs.sentry.io/platforms/native/guides/breakpad/) to be [the successor](https://groups.google.com/a/chromium.org/g/chromium-dev/c/6eouc7q2j_g) to breakpad. The integration is straight forward. You include the header, create a client, and the library spawns the process for you. Crashes are saved in the [minidump](https://docs.sentry.io/platforms/native/guides/minidumps/#what-is-a-minidump) format.

I've integrated [crashpad] for Windows on my [`feat/crashpad` branch](https://github.com/Nerixyz/chatterino2/tree/feat/crashpad). This requires one to get crashpad from [conan](https://conan.io/center/crashpad), since it doesn't have a CMake integration (we could use [sentry-crashpad](https://github.com/getsentry/crashpad) which is a fork that has CMake integration). Crashes are placed into `<chatterino-dir>/Crashes`. Chatterino is built in `RelWithDebInfo` mode. On my end, this inflates the size from about 8 MiB to about 16 MiB.

### Crash analysis

To analyze crashes, you need a symbol file of the executable (or any other debug information) and a minidump. Then you can use tools like [`minidump-stackwalk`](https://lib.rs/crates/minidump-stackwalk) to view the information. On Windows, you can also use [WinDbg](https://www.microsoft.com/store/p/windbg/9pgjgd53tn86) to analyze.

### Example

I've intentionally dereferenced a null pointer.
I built chatterino like this:

```powershell
mkdir build && cd build
conan install .. -s build_type=RelWithDebInfo -b missing -pr:b=default
cmake -DCMAKE_TOOLCHAIN_FILE=conan_toolchain.cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo -DBUILD_WITH_CRASHPAD=On -GNinja ..
cmake --build .
cmake --build . --target crashpad_handler
windeployqt --release --no-compiler-runtime --no-translations --no-opengl-sw --dir bin/ .\bin\chatterino.exe
```

Then I extracted the symbol information using [dump_syms](https://lib.rs/crates/dump_syms):
```powershell
cd bin
dump_syms chatterino.exe -s store
7z a symbols.7z store/*
```
[symbols.zip](https://github.com/Chatterino/chatterino2/files/10549430/symbols.zip)

Now I ran into the crash. A `.dmp` file was generated in `Crashes/reports` ([crash.zip](https://github.com/Chatterino/chatterino2/files/10549398/crash.zip)).

Now I can analyze the crash:

```powershell
> minidump-stackwalk .\Crashes\reports\0da3bb92-1d13-40e8-9267-0aa13810aebd.dmp --symbols-path store | head -n 30
Operating system: Windows NT
                  10.0.19045 2486
CPU: amd64
     family 6 model 60 stepping 3
     8 CPUs

Crash reason:  EXCEPTION_ACCESS_VIOLATION_READ
Crash address: 0x8
Process uptime: 33 seconds

Thread 0  (crashed)
 0  chatterino.exe!chatterino::TwitchChannel::sendMessage(QString const&) [TwitchChannel.cpp : 406 + 0x2e]
     rax = 0x000000495d0f3368    rdx = 0x0000000000000000
     rcx = 0x000001a97bb81cf0    rbx = 0x000000495d0f7290
     rsi = 0x000000495d0f3530    rdi = 0x000000495d0f3460
     rbp = 0x000000495d0f3399    rsp = 0x000000495d0f3320
      r8 = 0x000001a902a72e90     r9 = 0x0000000000000040
     r10 = 0x00000000000000c0    r11 = 0x018a073660142580
     r12 = 0x0000000000000003    r13 = 0x000001a90002f6f0
     r14 = 0x000001a9000bf4d0    r15 = 0x000000495d0f34e8
```

And indeed, line 406 is the line I dereferenced a null-pointer.

[crashpad]: https://chromium.googlesource.com/crashpad/crashpad
[breakpad]: https://chromium.googlesource.com/breakpad/breakpad/

**Comment by pajlada** [comment]
Sounds good to me, a big blocker for implementing breakpad fully was the mess our build system was in at the time. Now, being cmake only, this seems like a good path forward.

I'm happy to have crashes just be stored in a local directory with no upload stage as a V1 implementation.

I'm also ok for the implementation to be done one platform at a time, so if you have a version that you know works for Windows, just ifdef it to windows and have it live in master for a while.

**Comment by Felanbird** [comment]
Added in https://github.com/Chatterino/chatterino2/pull/4351
